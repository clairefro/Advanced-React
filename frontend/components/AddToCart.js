import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Link from 'next/link';
import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

class AddToCart extends Component {
  render() {
    const { id, me } = this.props;
    return (
      <Mutation
        mutation={ADD_TO_CART_MUTATION}
        variables={{ id }}
        refetchQueries={[{query: CURRENT_USER_QUERY}]}>
      {(addToCart, { loading }) => {
        if(!me) return <Link href={{pathname:'/signup'}}><a>Sign in to add to cart</a></Link>;
        return (
          <button disabled={loading} onClick={addToCart}>
            <i className="las la-plus"></i>Add{loading && 'ing'} to Cart
          </button>
        );
      }}
      </Mutation>
    );
  }

}

export default AddToCart;
export { ADD_TO_CART_MUTATION };
