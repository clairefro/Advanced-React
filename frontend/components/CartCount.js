import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { TransitionGroup, CSSTransition } from 'react-transition-group';

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(4) rotateX(0.5turn) skew(-5deg) rotate(-1deg);
  }
  .count-enter-active {
    transform: rotateX(0) skew(-5deg) rotate(-1deg);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: scale(4) rotateX(0) skew(-5deg) rotate(-1deg);
  }
  .count-exit {
    transform: rotateX(0.5turn) skew(-5deg) rotate(-1deg);
  }

`;

const Dot = styled.div`
  background: ${props => props.theme.red};
  color: white;
  transform: skew(-5deg) rotate(-1deg);
  padding: 0.5rem;
  linheight: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: 'tnum';
  font-varient-numeric: tabular-nums;
`;

const CartCount = ({ count }) =>  (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}>
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>

);

export default CartCount;
