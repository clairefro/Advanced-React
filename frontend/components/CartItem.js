import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Link from 'next/link';
import RemoveFromCart from './RemoveFromCart';
import formatMoney from '../lib/formatMoney';

const CartItemStyles = styled.li`
  a {
    padding: 1rem 0;
    border-bottom: 1px solid ${props => props.theme.lightgrey};
    display: grid;
    align-items: center;
    grid-template-columns: auto 1fr auto;
    img {
      margin-right: 10px;
    }
    h3,p {
      margin: 0;
    }
  }
`;

const CartItem = ({ cartItem }) => {
  // first check if cart item still exists (wasn't deleted)
  if(!cartItem.item) return (
      <CartItemStyles>
      <p>This item is no longer available</p>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  );
  return (
    <CartItemStyles>
    <Link href={{
      pathname: '/item',
      query: { id: cartItem.item.id}
    }}>
       <a>
         <img
           src={cartItem.item.image}
           alt={cartItem.item.title}
           width="100px"
           />
         <div className="cart-item-details">
           <h3>{cartItem.item.title}</h3>
           <p>
             {formatMoney(cartItem.item.price * cartItem.quantity)}
             {' - '}
             <em>
               {cartItem.quantity} &times; {formatMoney(cartItem.item.price)} ea.
             </em>
             </p>
         </div>
         <RemoveFromCart id={cartItem.id} />
       </a>
      </Link>
    </CartItemStyles>
  );
}


CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired,
}

export default CartItem;
