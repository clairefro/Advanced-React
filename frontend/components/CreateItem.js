import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import formatMoney from '../lib/formatMoney';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
    isLoadingImage: false,
  }

  handleChange = (e) => {
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;

    // name is 'title', 'price', 'description' etc
    this.setState({ [name]: val });
  }

  uploadFile = async (e) => {
    // console.log('uploading file...');
    // get file
    this.setState({ isLoadingImage: true });
    const files = e.target.files;
    const data = new FormData();
    // these below are needed by cloudinary
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');
    // hit the cloudinary api
    const res = await fetch('https://api.cloudinary.com/v1_1/dgxkozw6v/image/upload', {
        method: "POST",
        body: data
    });
    const file = await res.json();
    // console.log(file);

    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
      isLoadingImage: false,
    });
  }

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (

        <Form data-test="form" onSubmit={async e => {
          e.preventDefault();
          // push data from state to server to create item
          const res = await createItem();
          // console.log(res);

          // redirect to show page
          Router.push({
            pathname: '/item',
            query: { id: res.data.createItem.id }
          });

        }}>
          <ErrorMessage error={error}/>
          <fieldset disabled={loading} aria-busy={loading || this.state.isLoadingImage}>

            <label htmlFor="file">
              Socks Image
              <input
                type="file"
                id="file"
                name="file"
                placeholder="Upload an image"
                required
                onChange={this.uploadFile}
              />
              {this.state.isLoadingImage && <p className="loading-text">Loading your image...</p>}
              {this.state.image &&
                <img width="200"
                     src={this.state.image
                         .replace('https://res.cloudinary.com/advreactwgqlwb/image/upload/',
                                  'https://res.cloudinary.com/advreactwgqlwb/image/upload/t_media_lib_thumb/'
                         )}
                     alt="preview"/>}
            </label>

            <label htmlFor="title">
              Title
              <input
                type="text"
                id="title"
                name="title"
                placeholder="Title"
                required
                value={this.state.title}
                onChange={this.handleChange}
              />
            </label>

            <label htmlFor="price">
              Price in cents (USD)
              <input
                type="number"
                id="price"
                name="price"
                placeholder="Price"
                required
                value={this.state.price}
                onChange={this.handleChange}
              />
            </label>

            <label htmlFor="description">
              Description
              <input
                id="description"
                name="description"
                placeholder="Enter a description"
                required
                value={this.state.description}
                onChange={this.handleChange}
              />
            </label>
            <button disabled={this.state.isLoadingImage ? true : false} type="submit">Sell them!</button>
          </fieldset>
        </Form>

      )}
    </Mutation>
    );
  }

}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
