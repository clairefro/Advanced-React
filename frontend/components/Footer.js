import React from 'react';
import Link from 'next/link';
import styled from 'styled-components';
import Router from 'next/router';

const FooterStyles = styled.div`
  background-color: ${props => props.theme.grey};
  color: ${props => props.theme.offWhite};
  padding: 20px;
  margin-top: 60px;
  flex-shrink: 0;
  /* position: absolute; */
  /* bottom: 0; */
  width: 100%;
  height: 20rem;
  justify-self: flex-end;
`;

const IconLinks = styled.div`
  font-size: 3rem;
  margin-right: 1rem;
  a {
    color: ${props => props.theme.offWhite};
    margin-right: 0.5rem;
    &:hover {
      color: ${props => props.theme.red};
    }
  }
`;

const Footer = (props) => (
  <FooterStyles>
    <p>TODO: Write a footer</p>
    <IconLinks>
      <a href="#"><i className="lab la-instagram"></i></a>
      <a href="#"><i className="lab la-facebook-square"></i></a>
      <a href="#"><i className="lab la-twitter"></i></a>
      <a href="#"><i className="lab la-vimeo"></i></a>
    </IconLinks>
  </FooterStyles>
);

export default Footer;
