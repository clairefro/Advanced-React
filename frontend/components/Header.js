import React from 'react';
import Link from 'next/link';
import styled from 'styled-components';
import Router from 'next/router';
import NProgress from 'nprogress';

import Nav from './Nav';
import Cart from './Cart';
import Search from './Search';

Router.onRouteChangeStart = () => {
  NProgress.start();
}
Router.onRouteChangeComplete = () => {
  NProgress.done();
}
Router.onRouteChangeError = () => {
  NProgress.done();
}


const Logo = styled.h1`
  font-size: 2rem;
  margin-left: 2rem;
  position: relative;
  padding: 1rem 0;
  z-index: 2;
  transform: skew(-7deg);
  a {
    padding: 0.5rem 1.5rem;
    background: ${props => props.theme.red};
    color: white;
    text-transform: uppercase;
    text-decoration: none;
  }
  img {
    transform: skew(21deg);
    height: 50px;
    postion: relative;
    right: -5px;
  }
  @media (max-width: 1300px) {
    text-align: center;
    margin: 0;
  }
`;

const StyledHeader = styled.div`
  .bar {
    border-bottom: 10px solid ${props => props.theme.black};
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
    @media (max-width: 1300px) {
      grid-template-columns: 1fr;
      justify-content: center;
    }
  }
  .subbar {
    display: grid;
    grid-template-columns: 1fr auto;
    border-bottom: 1px solid ${props => props.theme.lightgrey};
  }
`;


const Header = (props) => (
  <StyledHeader>
    <div className='bar'>
      <Logo>
        <img src="../static/favicon.png" alt="nyse-logo"/>
        <Link href="/">
          <a>NY Sock Exchange</a>
        </Link>
      </Logo>
      <Nav />
    </div>
    <div className='subbar'>
      <Search />
    </div>
    <Cart />
  </StyledHeader>
);

export default Header;
