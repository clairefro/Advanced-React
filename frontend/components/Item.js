import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';

import Title from './styles/Title';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';
import Error from './ErrorMessage';
import User from './User';
import formatMoney from '../lib/formatMoney'

class Item extends Component {
  static propTypes = {
    item: PropTypes.object.isRequired,
  };

  render() {
    const { item } = this.props;
    return (
      <User>
      {({ data: { me }})=> {
        return (
          <ItemStyles>
          <Link href={{
            pathname: '/item',
            query: { id: item.id },
          }}>
            <a>
              {item.image && <img src={item.image} alt={item.title}/>}
              <Title>
              <a>{item.title}</a>
              </Title>
              <PriceTag>{formatMoney(item.price)}</PriceTag>
              <p>{item.description}</p>
            </a>
          </Link>
          <div className="buttonList">
          {me && me.id === item.user.id &&
            <Link href={{
              pathname: 'update',
              query: { id: item.id },
            }}>
              <a>Edit</a>
            </Link>
          }
          <AddToCart id={item.id} me={me}/>
          {me && me.id === item.user.id &&
          <DeleteItem id={item.id}>Delete</DeleteItem>
          }
          </div>
          </ItemStyles>
        );
      }}
      </User>

    );
  }

}

export default Item;
