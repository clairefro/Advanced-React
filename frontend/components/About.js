import React, { Component } from 'react';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Link from 'next/link';
import { Query } from 'react-apollo';
import Head from 'next/head';
import OrderItemStyles from '../components/styles/OrderItemStyles';
import User from '../components/User';
import CartItem from '../components/CartItem';
import formatMoney from '../lib/formatMoney';

const USER_ITEMS_QUERY = gql`
  query USER_ITEMS_QUERY($id: ID!) {
    userItems(id: $id) {
      id
      image
      title
      price
      description
    }
  }
`;

const AccountDetailStyles = styled.li`
  box-shadow: ${props => props.theme.bs};
  list-style: none;
  width: 60vw;
  padding: 2rem;
  border-bottom: 4px solid ${props => props.theme.red};
  h2 {
    border-bottom: 2px solid red;
    margin-top: 0;
    margin-bottom: 2rem;
    padding-bottom: 2rem;
  }
  .link {
    color: ${props => props.theme.red};
    text-decoration: underline;
  }
`;

const ItemStyles = styled.div`
  a {
    padding: 1rem 0;
    border-bottom: 1px solid ${props => props.theme.lightgrey};
    display: grid;
    align-items: center;
    grid-template-columns: auto 1fr auto;
    img {
      margin-right: 2rem;
    }
    h3,p {
      margin: 0;
    }
  }
`;

const About = (props) => (
  <User>
  {({ data: { me }})=> (
    <AccountDetailStyles>
      <h2>Your account details</h2>
      <p>
        <span>Name:{' '}</span>
        <span>{me.name}</span>
      </p>
      <p>
        <span>Email:{' '}</span>
        <span>{me.email}</span>
      </p>
      <h2>Your socks for sale</h2>
      <Query query={USER_ITEMS_QUERY} variables={{ id: me.id }}>
      {(data)=> {
        if(!data.data.userItems.length) return <p>
        You haven't listed any socks yet...
        <Link href={{
          pathname: '/sell'
        }}>
        <a className="link">{' '}Sell some!</a>
        </Link></p>
        return data.data.userItems.map((item)=> (
          <ItemStyles key={item.id}>
            <Link href={{
              pathname: '/item',
              query: { id: item.id}
            }}>
               <a>
                 <img
                   src={item.image}
                   alt={item.title}
                   width="100px"
                   />
                 <div className="cart-item-details">
                   <h3>{item.title}</h3>
                   <p>
                     {formatMoney(item.price)}
                    </p>
                 </div>
               </a>
              </Link>
          </ItemStyles>
        ))
      }}
      </Query>
    </AccountDetailStyles>
  )}
  </User>
);

export default About;
