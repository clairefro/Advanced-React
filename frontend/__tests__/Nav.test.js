import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: {
      ...fakeUser(),
      cart: [fakeCartItem(),fakeCartItem(),fakeCartItem()]
    } } },
  },
];

describe("<Nav />", () => {
  it('renders a minimal nav when not signed in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // console.log(wrapper.debug());
    const nav = wrapper.find('[data-test="nav"]');
    // console.log(toJSON(nav));
    // expect(toJSON(nav)).toMatchSnapshot();
    expect(nav.children().length).toBe(4);
  });

  it('renders a full nav when signed in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // console.log(wrapper.debug());
    // add ul selector because this data attribute is doubly rendered due to styled components
    const nav = wrapper.find('ul[data-test="nav"]');
    // console.log(toJSON(nav));
    // console.log(nav.debug());
    // expect(toJSON(nav)).toMatchSnapshot();
    expect(nav.children().length).toBe(6);
    expect(nav.text()).toContain('Sign out');
  });

  it('renders amount of items in cart', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocksWithCartItems}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // console.log(wrapper.debug());
    // add ul selector because this data attribute is doubly rendered due to styled components
    const nav = wrapper.find('ul[data-test="nav"]');
    const count = nav.find('div.count');
    // console.log(toJSON(nav));
    // console.log(nav.debug());
    // expect(toJSON(nav)).toMatchSnapshot();
    expect(toJSON(count)).toMatchSnapshot();
  });
});
