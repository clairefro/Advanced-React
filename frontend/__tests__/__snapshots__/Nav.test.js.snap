// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders amount of items in cart 1`] = `
<div
  className="CartCount__Dot-xxvp4g-1 gpIeEZ count"
>
  9
</div>
`;
